fuzz-tests/ntt.cpp

```diff
@@ -21,49 +21,52 @@ ll modpow(ll a, ll e) {
 }
 
 typedef vector<ll> vl;
-void ntt(ll* x, ll* temp, ll* roots, int N, int skip) {
-	if (N == 1) return;
-	int n2 = N/2;
-	ntt(x     , temp, roots, n2, skip*2);
-	ntt(x+skip, temp, roots, n2, skip*2);
-	rep(i,0,N) temp[i] = x[i*skip];
-	rep(i,0,n2) {
-		ll s = temp[2*i], t = temp[2*i+1] * roots[skip*i];
-		x[skip*i] = (s + t) % mod; x[skip*(i+n2)] = (s - t) % mod;
-	}
-}
-void ntt(vl& x, bool inv = false) {
-	ll e = modpow(root, (mod-1) / sz(x));
-	if (inv) e = modpow(e, mod-2);
-	vl roots(sz(x), 1), temp = roots;
-	rep(i,1,sz(x)) roots[i] = roots[i-1] * e % mod;
-	ntt(&x[0], &temp[0], &roots[0], sz(x), 1);
+int mul(int a, int b) { return (ll)a * b % mod; }
+int sub(int a, int b) { return b > a ? a - b + mod : a - b; }
+int add(int a, int b) { return a + b >= mod ? a + b - mod : a + b; }
+void ntt(vi &a, vi &rt, vi &rev, int n) {
+    rep(i, 0, n) if (i < rev[i]) swap(a[i], a[rev[i]]);
+    for (int k = 1; k < n; k *= 2)
+        for (int i = 0; i < n; i += 2 * k)
+            rep(j, 0, k) {
+                int z = mul(rt[j + k], a[i + j + k]);
+                a[i + j + k] = sub(a[i + j], z);
+                a[i + j] = add(a[i + j], z);
```

Comment:
Did this improve performance much over the previous version? This requires inputs to be within [0, mod), and adds add/sub functions.

Reply:
https://ideone.com/emwmZA

About 25% gain. We can inline the add/sub functions.